recourse_methods/test_dice_method.py

```diff
@@ -0,0 +1,118 @@
+            dice_kwargs=mock_dice_kwargs,
+            dice_counterfactual_kwargs=mock_counterfactual_kwargs,
+        )
+
+        expected_counterfactual_kwargs = mock_counterfactual_kwargs.copy()
+        expected_counterfactual_kwargs["stopping_threshold"] = mock_confidence
+        self.assertEqual(
+            expected_counterfactual_kwargs, dice.dice_counterfactual_kwargs
+        )
+
+        expected_dice_args = {
+            "data_interface": mock_dice_data(),
+            "model_interface": mock_model.to_dice_model(),
+        }
+        expected_dice_args.update(mock_dice_kwargs)
+        mock_dice.assert_called_with(**expected_dice_args)
+
+    @mock.patch(
+        "recourse_methods.dice_method.recourse_adapter.RecourseAdapter",
+        autospec=True,
+    )
+    def test_counterfactuals_to_directions(self, mock_adapter):
+        mock_adapter.transform.side_effect = lambda df: df * 10
+        mock_adapter.transform_series.side_effect = lambda series: series * 10
+        mock_self = mock.Mock(spec=dice_method.DiCE)
+        mock_self.adapter = mock_adapter
+
+        poi = pd.Series([0, 1], index=["col1", "col2"])
+        counterfactuals = pd.DataFrame(
+            {"col1": [5, -2], "col2": [3, 2]}, index=[10, 23]
+        )
+        expected_directions = pd.DataFrame(
+            {
+                "col1": [50 - 0, -20 - 0],
+                "col2": [30 - 10, 20 - 10],
+            }
+        )
+
+        directions = dice_method.DiCE._counterfactuals_to_directions(
+            mock_self, poi, counterfactuals
+        )
+
+        np.testing.assert_equal(
+            directions.to_numpy(), expected_directions.to_numpy()
+        )
+
+    @mock.patch("dice_ml.Dice", autospec=True)
+    @mock.patch("data.recourse_adapter.RecourseAdapter", autospec=True)
+    def test_generate_counterfactuals(
+        self,
+        mock_adapter: mock.Mock,
+        mock_dice: mock.Mock,
+    ):
+        mock_counterfactual_kwargs = {"mock_arg": "mock_val"}
+        mock_self = mock.Mock(spec=dice_method.DiCE)
+        mock_self.adapter = mock_adapter
+        mock_adapter.label_column = "mock_label"
+        mock_adapter.positive_label = "mock_value"
+        mock_self.dice = mock_dice
+        mock_self.dice_counterfactual_kwargs = mock_counterfactual_kwargs
+        poi = pd.Series([0, 0], index=["col1", "col2"])
+        num_counterfactuals = 2
+
+        _ = dice_method.DiCE._generate_counterfactuals(
+            mock_self, poi, num_counterfactuals
+        )
+        args = mock_dice.generate_counterfactuals.call_args.kwargs
+
+        expected_args = {
+            "query_instances": pd.DataFrame({"col1": [0], "col2": [0]}),
+            "total_CFs": 2,
+            "desired_class": 1,
+            "verbose": False,
+        }
+        expected_args.update(mock_counterfactual_kwargs)
+
+        # Check that DICE is called with correct arguments.
+        self.assertEqual(set(expected_args.keys()), set(args.keys()))
```

Comment:
This type of check seems better suited to using exception handling. Unclear if this is a useful unittest, but ensuring that the correct arguments are provided is useful. Maybe I'm misunderstanding something here?

Reply:
One thing you can be testing for here is that the counterfactual_kwargs get updated _if_ the self.dice_counterfactual_kwargs attribute already exists _and_ you have provided new values for the dictionary. If so, it'd be useful to make that clear in the test.

Reply:
This unit test only checks to make sure that the aguments passed to dice.generate_counterfactuals are formatted correctly and contain all the necessary info. I moved the formatting to a separate function so I can unit test it directly instead of through _generate_counterfactuals.